rpc/client/client.go

```diff
@@ -432,7 +438,65 @@ func (c *Client) TxSearch(ctx context.Context, query string, prove bool, pagePtr
 // BlockSearch defines a method to search for a paginated set of blocks by
 // BeginBlock and EndBlock event search criteria.
 func (c *Client) BlockSearch(ctx context.Context, query string, page, perPage *int, orderBy string) (*ctypes.ResultBlockSearch, error) {
-	panic("BlockSearch - not implemented!")
+	q, err := tmquery.New(query)
+	if err != nil {
+		return nil, err
+	}
+
+	results, err := c.node.BlockIndexer.Search(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort the results
+	switch orderBy {
+	case "desc":
+		sort.Slice(results, func(i, j int) bool {
+			return results[i] > results[j]
+		})
+
+	case "asc", "":
+		sort.Slice(results, func(i, j int) bool {
+			return results[i] < results[j]
+		})
+	default:
+		return nil, errors.New("expected order_by to be either `asc` or `desc` or empty")
+	}
+
+	// Fetch the blocks
+	var blocks []*optimint.Block
+	for _, h := range results {
+		block, err := c.node.Store.LoadBlock(uint64(h))
+		if err != nil {
+			return nil, err
+		}
+		blocks = append(blocks, block)
+	}
+
+	// Paginate
+	totalCount := len(results)
+	perPageVal := validatePerPage(perPage)
+
+	pageVal, err := validatePage(page, perPageVal, totalCount)
+	if err != nil {
+		return nil, err
+	}
+
+	skipCount := validateSkipCount(pageVal, perPageVal)
+	pageSize := tmmath.MinInt(perPageVal, totalCount-skipCount)
+
+	apiResults := make([]*ctypes.ResultBlock, 0, pageSize)
+	for i := skipCount; i < skipCount+pageSize; i++ {
```

Comment:
Lets call `LoadBlock` after pagination, only for selected blocks. This way we can reduce number of IO operations.

Reply:
Sure, I'm going with this approach 👍